src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -319,19 +340,22 @@ TransactionManager::do_submit_transaction(
   SUBTRACET(seastore_t, "start", tref);
   return trans_intr::make_interruptible(
     tref.get_handle().enter(write_pipeline.ool_writes)
-  ).then_interruptible([this, FNAME, &tref,
+  ).then_interruptible([this, &tref,
 			dispatch_result = std::move(dispatch_result)] {
     return seastar::do_with(std::move(dispatch_result),
-			    [this, FNAME, &tref](auto &dispatch_result) {
+			    [this, &tref](auto &dispatch_result) {
       return epm->write_delayed_ool_extents(tref, dispatch_result.alloc_map
-      ).si_then([this, FNAME, &tref, &dispatch_result] {
-        SUBTRACET(seastore_t, "update delayed extent mappings", tref);
-        return lba_manager->update_mappings(tref, dispatch_result.delayed_extents);
-      }).handle_error_interruptible(
+      ).handle_error_interruptible(
         crimson::ct_error::input_output_error::pass_further(),
         crimson::ct_error::assert_all("invalid error")
       );
     });
+  }).si_then([&tref, this] {
+    return tref.get_handle().enter(write_pipeline.update_lba_mappings);
+  }).si_then([&tref, this] {
+    return update_lba_mappings(tref);
```

Comment:
I suppose traversing the lba-tree (again) is expensive. Is it possible to instead leverage the lba parent-child-pointer to update the mapping of the logical extent?

Reply:
> I suppose traversing the lba-tree (again) is expensive. Is it possible to instead leverage the lba parent-child-pointer to update the mapping of the logical extent?

This may be non-trivial, because the lba tree structure may be changed after the retrieval of an lba mapping, so we might not be able to update the mapping directly.

On the other hand, we already are updating the lba mappings for the written ool extents here. So I think this may not be so much increased costs, am I right?

Reply:
> This may be non-trivial, because the lba tree structure may be changed after the retrieval of an lba mapping, so we might not be able to update the mapping directly through the LBAMapping structure. To archieve this, we may have to expose the lba tree iterator somehow?

Maybe I'm missing something. Quoted:
```
3. if nodes are initial_pending
    a. parent points at its pending parent on this transaction (must exist)
    b. prior_instance is empty or, if it's the result of rewrite, points to
        its stable predecessor
    ...
    d. copy_sources contains the set of stable nodes at the same tree-level(only
       its "prior_instance" if the node is the result of a rewrite), with which
       the lba range of this node overlaps.
```
Seems to me this design is already sufficient to find the parent node, which is always kept updated with tree split/merge, right?

Or, does the relationship between LBA leaf node and the child LogicalExtent follow the above property?

> On the other hand, we already are updating the lba mappings for the written ool extents here. The major difference is that we update the checksum field of lba_map_val_t in addition to the paddr field, so I think there may not be so much increased costs, am I right?

If the above method makes sense, it should apply to the update_mapping path generally, including updating the paddr field.

Reply:
> > This may be non-trivial, because the lba tree structure may be changed after the retrieval of an lba mapping, so we might not be able to update the mapping directly through the LBAMapping structure. To archieve this, we may have to expose the lba tree iterator somehow?
> 
> Maybe I'm missing something. Quoted:
> 
> ```
> 3. if nodes are initial_pending
>     a. parent points at its pending parent on this transaction (must exist)
>     b. prior_instance is empty or, if it's the result of rewrite, points to
>         its stable predecessor
>     ...
>     d. copy_sources contains the set of stable nodes at the same tree-level(only
>        its "prior_instance" if the node is the result of a rewrite), with which
>        the lba range of this node overlaps.
> ```
> 
> Seems to me this design is already sufficient to find the parent node, which is always kept updated with tree split/merge, right?

So you're suggesting that update_mapping go backward to the parent when the current leaf node doesn't contain a specific key? If so, I think it's feasible.

> 
> Or, does the relationship between LBA leaf node and the child LogicalExtent follow the above property?

IIRC, it follows the property.

> 
> > On the other hand, we already are updating the lba mappings for the written ool extents here. The major difference is that we update the checksum field of lba_map_val_t in addition to the paddr field, so I think there may not be so much increased costs, am I right?
> 
> If the above method makes sense, it should apply to the update_mapping path generally, including updating the paddr field.

I think we can implement this in a dedicated lba mapping update perf optimization PR.

Reply:
Per discussion, the optimization relies on the ability for a child (`LogicalCachedExtent`) to identify its direct parent (`LBALeafNode`) within the transaction.

The missing piece is that if the parent is initial-pending and the child is stable or mutation-pending, the child cannot identify the parent because the pointers from an initial-pending extent to its stable version is currently only one-directional though its `copy_sources`.